Fix checkpoint record corrupt with gp_before_filespace_setup

yaowangm opened the pull request:
### Issue

When we turned on the GUC `gp_before_filespace_setup` and restarted GPDB, we hit an assertion in `UnpackCheckPointRecord()`:

```
8393         Assert(remainderLen > ckptExtended->masterMirroringCheckpointLen);
```
Relevant stack:
```
#0  0x00007f47d42406b3 in __select_nocancel () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000c03b8d in pg_usleep (microsec=30000000) at pgsleep.c:43
#2  0x0000000000a727dd in elog_debug_linger (edata=0x11f4940 <errordata>) at elog.c:4342
#3  0x0000000000a69d31 in errfinish (dummy=0) at elog.c:616
#4  0x0000000000a6847c in ExceptionalCondition (conditionName=0xc1f0d0 "!(remainderLen > ckptExtended->masterMirroringCheckpointLen)", errorType=0xc1a9ea "FailedAssertion", fileName=0xc1aa32 "xlog.c", lineNumber=8393) at assert.c:49
#5  0x000000000051a6ce in UnpackCheckPointRecord (record=0x7f47d5a72010, ckptExtended=0x7ffc642fc710) at xlog.c:8393
#6  0x0000000000516dd9 in StartupXLOG () at xlog.c:6685
#7  0x00000000005202fa in StartupProcessMain (passNum=1) at xlog.c:11270
#8  0x000000000059e8bc in AuxiliaryProcessMain (argc=2, argv=0x7ffc642fc840) at bootstrap.c:462
#9  0x00000000008b6601 in StartChildProcess (type=StartupProcess) at postmaster.c:7804
#10 0x00000000008ac388 in StartMasterOrPrimaryPostmasterProcesses () at postmaster.c:1562
#11 0x00000000008b929b in stepDoDatabaseStartup (nextState=TSDone, stateInOut=0x7ffc642fc94c) at primary_mirror_mode.c:1738
#12 0x00000000008b975b in applyStepForTransitionToMasterOrMirrorlessMode (args=0x7ffc642fc960, stateInOut=0x7ffc642fc94c, extraResultInfoOut=0x7ffc642fcb20 "") at primary_mirror_mode.c:1907
#13 0x00000000008b8e71 in doRequestedPrimaryMirrorModeTransitions (isInShutdown=0 '\000') at primary_mirror_mode.c:1643
#14 0x00000000008ad738 in ServerLoop () at postmaster.c:2268
#15 0x00000000008ac323 in PostmasterMain (argc=14, argv=0x2e35a30) at postmaster.c:1528
#16 0x00000000007ba4f3 in main (argc=14, argv=0x2e35a30) at main.c:206
```
In a production env with assertion disabled, we even hit segment fault when reading checkpoint record.

### Analysis

In a brief, the root cause is, with `gp_before_filespace_setup`=on, `CreateCheckPoint()` created a corrupt checkpoint xlog record which didn't contain mmxlog but did contain `ptas` (prepared transaction aggregate state),  and `UnpackCheckPointRecord()` cannot handle it.

The wrong version was introduced by https://github.com/greenplum-db/gpdb/pull/9730. 

Something interesting is, the description of `gp_before_filespace_setup` is "Indicates that the gp_persistent_filespace_node table is not setup and should not be used for lookups." So a reasonable understanding is that "with gp_before_filespace_setup mmxlog will be skipped but ptas will not". I was believing it, and I suppose the author of PR #9730 also believed it. However, `UnpackCheckPointRecord()` always expect no `ptas` with `gp_before_filespace_setup`:
```
 8378     /*
 8379      * The master mirror checkpoint (mmxlog) and prepared transaction aggregate state (ptas) will be skipped
 8380      * when gp_before_filespace_setup is ON.
 8381      */
```

See the different rdata chain generated by wrong (after PR #9730) and correct code (before PR #9730) with `gp_before_filespace_setup`=on:

Wrong version:
```
(gdb) p rdata[0]
$1 = {data = 0x7fff8b2d8840 "", len = 48, buffer = 0, buffer_std = -64 '\300', next = 0x7fff8b2d88c0}
(gdb) p rdata[1]
$2 = {data = 0x562d9295d890 "", len = 4, buffer = 0, buffer_std = 4 '\004', next = 0x7fff8b2d8940}
...
(gdb) p &rdata[5]
$12 = (XLogRecData *) 0x7fff8b2d8940
```
Correct version:
```
(gdb) p rdata[0]
$2 = {data = 0x7ffebc62c510 "", len = 48, buffer = 0, buffer_std = 96 '`', next = 0x7ffebc62c560}
(gdb) p rdata[1]
$3 = {data = 0x55ce673feb70 "", len = 4, buffer = 0, buffer_std = 4 '\004', next = 0x0}
```
For correct version, `rdata[1].next` is null and the chain stopped here. For wrong version, `rdata[1].next` is 0x7fff8b2d8940, which is actually the address of `rdata[5]`.

Relevant code in `CreateCheckPoint()`:
```
 9071     rdata[1].buffer = InvalidBuffer;
 9072     rdata[1].next = NULL;
 9073     pnext = &rdata[1].next;
 ...
 9081     /* mmxlog_append_checkpoint_data() may use rdata[2,3,4] */
 9082     pnext = mmxlog_append_checkpoint_data(&rdata[2], pnext);
...
 9088     *pnext = &rdata[5];
 9089     rdata[5].data = (char*)p;
 9090     rdata[5].buffer = InvalidBuffer;
 9091     rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
 9092     rdata[4].next = &(rdata[5]);
 9093     rdata[5].next = NULL;
```
Before the calling of `mmxlog_append_checkpoint_data()` (line 9082), `pnext` pointed to `&rdata[1].next`.  

With `gp_before_filespace_setup`=off, `pnext` changed to `&rdata[4].next` inside `mmxlog_append_checkpoint_data()` and then was set to `&rdata[5]`, which generated a valid chain. 

With `gp_before_filespace_setup`=on,  `mmxlog_append_checkpoint_data()` returned earlier and `pnext` still pointed to `&rdata[1].next` after the calling, then pointed to `&rdata[5] `and generated an invalid chain. 

### Fix
The fix is straightforward: 

The pointer `pnext` is supposed to advance after getting mmxlog data inside `mmxlog_append_checkpoint_data()`. We check whether `pnext` has advanced after the calling: if not, it means mmxlog was not collected due to returning earlier, and the `rdata` chain stops here. If not, go ahead and link to `ptas` data.

A further thought is, can we avoid all the codes of collection `ptas` data when `gp_before_filespace_setup`=on? Or an reverse approach also makes sense (maybe better): updating `UnpackCheckPointRecord()` and other parsing codes to handle checkpoint xlog record without mmxlog but with `ptas` . I took a basic look and it seems relevant code is complex. Since GPDB5 is a deprecated version, I don't want to impose more change with potential risk. Let's restrict to the fix itself for now.

ashwinstar commented:
I would like the commit message to describe exactly the problem and the fix (too much is left for the reviewers to figure out :))

haolinw commented on src/backend/access/transam/xlog.c:
```diff
@@ -9085,11 +9085,32 @@ CreateCheckPoint(int flags)
+		/*
+		 * If pnext has advanced, gp_before_filespace_setup must be false.
+		 */
+		Assert(!gp_before_filespace_setup);
```
It seems to me necessary to add one more `Assert(pnext == &rdata[4].next);` as well to ensure and represent clearly current pnext value is where we are expecting.

yaowangm replied:
Good point. Done.

yaowangm commented on src/backend/access/transam/xlog.c:
```diff
@@ -9085,11 +9085,32 @@ CreateCheckPoint(int flags)
-	rdata[5].next = NULL;
+
+	/*
+	 * Note the master mirror checkpoint (mmxlog) and prepared transaction
```
Please see the PR description I added just now.

haolinw replied:
Is that necessary still calling getTwoPhasePreparedTransactionData() if gp_before_filespace_setup was set to on ?

yaowangm replied:
ooh, it seems a little bit more complex.

I did try to avoid getTwoPhasePreparedTransactionData() but failed, because the pointer `p` is stilled used later to calculater `ptrd_oldest_ptr` (see line 9161). So I gave up. However, after checking the code related to `ptrd_oldest_ptr`, it seems whether we add `ptas` or not will affect the later offset computing. I need a little bit more time to understand the relevant code to determine the correct behavior without `ptas`. I will update later.

yaowangm replied:
@haolinw  Updated code. As my understanding, with gp_before_filespace_setup=on we need to also avoid computing `ptrd_oldest_ptr` related code and follow different branch on line 9214 finally:

```
 9210     /*
 9211      * Select point at which we can truncate the log, which we base on the
 9212      * prior checkpoint's earliest info or the oldest prepared transaction xlog record's info.
 9213      */
 9214     if (ptrd_oldest_ptr != NULL && XLByteLE(ptrd_oldest, ControlFile->checkPointCopy.redo))
 9215         XLByteToSeg(ptrd_oldest, _logId, _logSeg);
 9216     else
 9217         XLByteToSeg(ControlFile->checkPointCopy.redo, _logId, _logSeg);
```

We can talk more about details tomorrow. Thanks for pointing it out.

haolinw replied:
Looks fine. I just tried to replay Ashwin's comments ;-)

yaowangm replied:
@haolinw As discussed today, I reverted the code to always collect ptas data since it must be used to calculate offset of deleting xlog files. I also have added detailed comments. Please take another look, thx.

haolinw replied:
LGTM, a minor suggestion is combing both comments together but separate with different segment might look better.
```
         .....
         * is ON. See comments inside UnpackCheckPointRecord(). For the case,
	 * mmxlog_append_checkpoint_data() will return earlier.
         *
	 * Note that we need to collect ptas data even when we don't need to write
	 * it to check point xlog record.
	 * When gp_before_filespace_setup is ON, we will not link mmxlog and ptas
	 * to rdata. However, we still need ptas data to calculate oldest prepared
	 * transaction XLogRecPtr (ptrd_oldest) and offset of deleting old log
	 * files. See the code related to ptrd_oldest/ptrd_oldest_ptr/_logId/
	 * _logSeg.
```

yaowangm replied:
Thanks. Updated.

yaowangm commented:
> I would like the commit message to describe exactly the problem and the fix (too much is left for the reviewers to figure out :))

As talked, I added detailed description and removed the 'WIP' tag. Please take another look, thx.

ashwinstar commented:
I didn't check, what was the logic before PR #9730? Was the prepared transaction record skipped as well from writing if `gp_before_filespace_setup` Understand `UnpackCheckPointRecord()` logic is written in that fashion but it may not be the right thing to do? We really need to make it based on whether can prepared transaction exists when `gp_before_filespace_setup` is set. If yes then that information has to be carried to the checkpoint as well even if mmxlog record information is being skipped due to guc. With our minimal test it was found ptas returned on prepared transactions hence it was fine to skip, but that may not be sufficient for all cases. So, finding the behavior before the regression PR #9730 would be really helpful.

yaowangm commented:
> I didn't check, what was the logic before PR #9730? Was the prepared transaction record skipped as well from writing if `gp_before_filespace_setup` Understand `UnpackCheckPointRecord()` logic is written in that fashion but it may not be the right thing to do? We really need to make it based on whether can prepared transaction exists when `gp_before_filespace_setup` is set. If yes then that information has to be carried to the checkpoint as well even if mmxlog record information is being skipped due to guc. With our minimal test it was found ptas returned on prepared transactions hence it was fine to skip, but that may not be sufficient for all cases. So, finding the behavior before the regression PR #9730 would be really helpful.

@ashwinstar Thanks for your comments.

As I mentioned in the description, I did check the logic before PR #9730 and compared it with the logic after PR #9730. In a brief, yes, **the old logic is to skip ptas as well** when `gp_before_filespace_setup` is on, so the fix is  just to back to the old logic. In fact, I was not able to find any fault in existing code since the generated chain looked good, and I found out the root cause only when I compared the old/new logic.

In addition, I also mentioned that `UnpackCheckPointRecord()` explicitly points out that ptas will be skipped for the case: 

 8378     /*
 8379      * The master mirror checkpoint (mmxlog) and prepared transaction aggregate state (ptas) will be skipped
 8380      * when gp_before_filespace_setup is ON.
 8381      */

So I suppose it is the design, unless we can prove the understanding of `UnpackCheckPointRecord()` is incorrect.

I also mentioned my thoughts about another approach:

> Or an reverse approach also makes sense (maybe better): updating UnpackCheckPointRecord() and other parsing codes to handle checkpoint xlog record without mmxlog but with ptas . I took a basic look and it seems relevant code is complex. Since GPDB5 is a deprecated version, I don't want to impose more change with potential risk. Let's restrict to the fix itself for now.

Please let me know if you have any more comments.

ashwinstar commented:
Thank you for the digging and explanation. Agree no need to make a functional change to the STABLE branch. Just having behavior back to the prior version of PR #9730 seems the perfect route to go.
